templates/load-message.lua

```diff
@@ -0,0 +1,57 @@
+-- Some parameters to control the behavior. Can be set via Redis keys.
+-- How long (in seconds) should a given buffer bucket be?
+local pBufferBucketInterval = tonumber(redis.call("GET", "conf-bufferBucketInterval")) or 60
+-- How many buffer buckets should we retain?
+local pBufferBucketCount = tonumber(redis.call("GET", "conf-bufferBucketCount")) or 20
+
+-- Derived parameters
+local pBufferLifetime = pBufferBucketInterval * pBufferBucketCount
+
+-- First, parse and restructure the incoming message.
+local message = cjson.decode(ARGV[1])
+message.redisMessageId = tonumber(redis.call("INCR", "last-redisMessageId"))
+local messageJson = cjson.encode(message)
+
+-- It's OK to index messages by app name, since app names are unique within an
+-- environment.
+local appName = message.app
+
+-- Now, make sure we deliver the message to subscribers waiting for it.
+local publishRet = redis.call("PUBLISH", "stream-" .. appName, messageJson)
+
+-- Now, we need to add the key to a buffer. We actually want to use multiple buffers
+-- for each log stream, so that the Redis volatile-ttl policy will automatically
+-- expire old logs for us, without entirely trimming the buffers. To that end, we
+-- buffer logs into a new bucket every minute, which means we'll have 10 active
+-- buckets per log stream, since our TTL on log buffers is 10 minutes.
+local bufferBucketIndex = math.floor(message.unix_timestamp / pBufferBucketInterval) * pBufferBucketInterval
+local bufferBucketName = "buffer-bucket-" .. appName .. "-" .. bufferBucketIndex
+
+-- New messages are added at the right end of the bucket, which means that
+-- messages will be sorted from first received to last received (note: the
+-- ordering within a bucket is based on the order in which the messages were
+-- received, not their actual timestamp).
+redis.call("RPUSH", bufferBucketName, messageJson)
+
+-- Set a  TTL, which ensures we eventually expire the bucket, and will be used
+-- by Redis to inform OOM eviction decisions (oldest buckets - which will
+-- expire the soonest - are evicted first).
+redis.call("EXPIREAT", bufferBucketName, message.unix_timestamp + pBufferLifetime)
+
+-- Finally, store the bucket name in a buffer map. The buffer map is a sorted
+-- set where we sort bucket names by their timestamp. Thus, when we query the
+-- set via ZRANGE ... 0 -1, we'll get the buckets from oldest to newest.
+local bufferMapName = "buffer-map-" .. appName
```

Comment:
@aaw I only briefly evaluated using `SCAN` (and `KEYS / related methods, etc.) before setting on the buffer map solution. There are a couple reasons why I did so

First, it's a little simpler for the consumer to extract the logs if we maintain a buffer map:

- We have a guarantee that the buffers will be in the right order when we query the buffer map, and that there won't be duplicates. Arguably, these are easy to work around in Dumptruck, but a little more annoying to work with in e.g. shell (though — granted — right now that's only used for tests).
- We have a guarantee that it only takes one request to Redis to retrieve a list of all available logs (and so it only takes 2 requests to get all the buffered logs out). 

Second, my understanding of `SCAN` is that it makes very little guarantees unless you're willing to query Redis an unbounded number of times to retrieve your records. Specifically, I'm referring to this note in the documentation:

> It is important to note that the MATCH filter is applied after elements are retrieved from the collection, just before returning data to the client. This means that if the pattern matches very little elements inside the collection, SCAN will likely return no elements in most iterations.

Ultimately, I feel like I better understand sets (and using `ZRANGE`), and I don't think there's such a high cost in maintaining the buffer map, so I'm more comfortable using it.

Let me know what you think!